articles/how-to-setup-user-authentication-from-scratch-with-rails-6/index.md

```diff
@@ -0,0 +1,702 @@
+end
+```
+
+- Remember to run your migrations.
+
+```bash
+rails db:migrate
+```
+
+Test our app,
+
+```bash
+rails server
+```
+
+We will see a Rails app with a home page.
+
+![localhost](/engineering-education/how-to-setup-user-authentication-from-scratch-with-rails-6/localhost.png)
+
+Test our model.
+
+```bash
+rails console
+```
+
+Create a new user
+
+![railsconsole](/engineering-education/how-to-setup-user-authentication-from-scratch-with-rails-6/railsconsole.png)
+
+### Configuring routes
+
+- Update `config/routes.rb`
+
+```rb
+Rails.application.routes.draw do
+  root 'welcome#index'
+
+  get 'sign_up', to: 'registrations#new'
+  post 'sign_up', to: 'registrations#create'
+
+  get 'sign_in', to: 'sessions#new'
+  post 'sign_in', to: 'sessions#create', as: 'log_in'
+
+  delete 'logout', to: 'sessions#destroy'
+
+  get 'password', to: 'passwords#edit', as: 'edit_password'
+  patch 'password', to: 'passwords#update'
+
+  get 'password/reset', to: 'password_resets#new'
+  post 'password/reset', to: 'password_resets#create'
+  get 'password/reset/edit', to: 'password_resets#edit'
+  patch 'password/reset/edit', to: 'password_resets#update'
+end
+```
+
+We are matching our routes to the above controller and actions
+
+### Adding Controllers
+
+Routers determine which controller to use for the request,controllers will then receive the request and save or fetch data from our models.
+
+- touch `app/controllers/registrations_controller.rb`
+
+```rb
+class RegistrationsController < ApplicationController
+  # instantiates new user
+  def new
+    @user = User.new
+  end
+
+  def create
+    @user = User.new(user_params)
+    if @user.save
+    # stores saved user id in a session
+      session[:user_id] = @user.id
+      redirect_to root_path, notice: 'Successfully created account'
+    else
+      render :new
+    end
+  end
+
+  private
+  def user_params
+    # strong parameters
+    params.require(:user).permit(:email, :password, :password_confirmation)
+  end
+end
+```
+
+`session` stores data for one request and used in another request. This controller is responsible for creating a new user and saving it to the database.
+
+- touch `app/controllers/sessions_controller.rb`
+
+```rb
+class SessionsController < ApplicationController
+  def new; end
+
+  def create
+    user = User.find_by(email: params[:email])
+    # finds existing user, checks to see if user can be authenticated
+    if user.present? && user.authenticate(params[:password])
+    # sets up user.id sessions
+      session[:user_id] = user.id
+      redirect_to root_path, notice: 'Logged in successfully'
+    else
+      flash.now[:alert] = 'Invalid email or password'
+      render :new
+    end
+  end
+
+  def destroy
+    # deletes user session
+    session[:user_id] = nil
+    redirect_to root_path, notice: 'Logged Out'
+  end
+end
+```
+`SessionsController` provides login functionality to an existing user, also logs out a user by deleting session data.
```

Comment:
More explanation required on each action in this controller. Please go through each and every line in all the actions.